worker/peergrouper/desired.go

```diff
@@ -103,22 +142,68 @@ func desiredPeerGroup(info *peerGroupInfo) ([]replicaset.Member, map[*machineTra
 	return memberSet, machineVoting, nil
 }
 
+// getLogMessage generates a nicely formatted log message for the input
+// replica-set arguments.
+func getLogMessage(members machineMembers, extra []replicaset.Member, maxId int) string {
+	lines := make([]string, 0)
+	lines = append(lines, fmt.Sprintf("calculated desired peer group\ndesired voting members: (maxId: %d)", maxId))
+	for tracker, replMem := range members {
+		lines = append(lines, fmt.Sprintf("\n   %#v: rs_id=%d, rs_addr=%s", tracker, replMem.Id, replMem.Address))
+	}
+	if len(extra) > 0 {
+		lines = append(lines, "\nother members:")
+		for _, replMem := range extra {
+			vote := replMem.Votes != nil && *replMem.Votes > 0
+			lines = append(lines, fmt.Sprintf("\n   rs_id=%d, rs_addr=%s, tags=%v, vote=%t",
+				replMem.Id, replMem.Address, replMem.Tags, vote))
+		}
+	}
+	return strings.Join(lines, "")
+}
+
+// checkExtraMembers checks to see if any of the input members, identified as
+// not being associated with machines, is set as a voter in the peer group.
+// If any have, an error is returned.
+// The boolean indicates whether any extra members were present at all.
+func checkExtraMembers(extra []replicaset.Member) (bool, error) {
+	for _, member := range extra {
+		if isVotingMember(&member) {
+			return false, fmt.Errorf("voting non-machine member %#v found in peer group", member)
+		}
+	}
+	return len(extra) > 0, nil
+}
+
 func isVotingMember(member *replicaset.Member) bool {
 	return member.Votes == nil || *member.Votes > 0
 }
 
-// possiblePeerGroupChanges returns a set of slices
-// classifying all the existing machines according to
-// how their vote might move.
-// toRemoveVote holds machines whose vote should
-// be removed; toAddVote holds machines which are
-// ready to vote; toKeep holds machines with no desired
-// change to their voting status (this includes machines
-// that are not yet represented in the peer group).
-func possiblePeerGroupChanges(
-	info *peerGroupInfo,
-	members map[*machineTracker]*replicaset.Member,
-) (toRemoveVote, toAddVote, toKeep []*machineTracker) {
+// getMongoAddresses gets an address suitable for Mongo peer group
+// communication for each tracked machine.
+// An error will be returned if more that one address is found for a machine
+// and there is no HA space is configured.
+func getMongoAddresses(info *peerGroupInfo) (
+	map[*machineTracker]string, error,
+) {
+	addrs := make(map[*machineTracker]string, len(info.machineTrackers))
+	for _, m := range info.machineTrackers {
+		var err error
+		if addrs[m], err = m.SelectMongoAddress(info.mongoPort, info.haSpace); err != nil {
```

Comment:
Prior to the change, the filtering by space/scope was in both _addNewMachines_ (for new members) and _updateAddresses_. I did the generation into this data structure so that:
- Determination is done once per machine.
- It is done early and we can error out from _desiredPeerGroup_ rather than having to check error returns from both methods later into the logic.
- It is passed to _updateAddresses_ to detect changes.

I will see if I can contrive something nicer for this.